lib/writable.js

```diff
@@ -1,69 +1,225 @@
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (!this.corked && this.writableLength) {
+      this.writing = true;
+      this.doWrite(this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    this.defaultEncoding = encoding;
+  }
+
+  pipe() {
+    this.emit('error', new Error('[ERR_STREAM_CANNOT_PIPE]'));
+  }
+
+  validChunk(chunk, cb) {
+    let error;
+    if (chunk === null) {
+      error = new Error('[ERR_STREAM_NULL_VALUES]');
+    } else if (typeof chunk !== 'string' && !Buffer.isBuffer(chunk)) {
+      error = new Error('[ERR_INVALID_ARG_TYPE]');
+    }
+
+    if (error) {
+      this.emit('error', error);
+      process.nextTick(cb, error);
+      return false;
+    }
+    return true;
+  }
+
+  write(chunk, encoding, cb) {
+    if (typeof encoding === 'function') {
+      cb = encoding;
+      encoding = null;
+    }
+
+    if (this.ended) {
+      const err = new Error('[ERR_STREAM_WRITE_AFTER_END]');
+      this.emit('error', err);
+      process.nextTick(cb, err);
+    }
+
+    if (this.destroyed) {
+      const err = new Error('[ERR_STREAM_DESTROYED]');
+      this.emit('error', err);
+      process.nextTick(cb, err);
+    }
+
+    if (!encoding) {
+      encoding = this.defaultEncoding;
+    }
+
+    if (typeof cb !== 'function') {
+      cb = () => {};
+    }
+
+    if (!this.validChunk(chunk, cb)) return;
+
+    if (this.corked || this.writing) {
+      this.buffer.add(chunk, encoding);
+      this.callbacks.push(cb);
+    } else {
+      if (typeof chunk === 'string') {
+        chunk = Buffer.from(chunk, encoding);
+      }
+      this.writing = true;
+      this.doWrite(chunk, [cb]);
+    }
+  }
+
+  doWrite(data, callbacks) {
+    if (!callbacks) {
+      callbacks = this.callbacks;
+      this.callbacks = [];
+    }
+
+    const cb = () => {
+      if (!this.writableLength) {
+        this.writing = false;
+        if (this.ending) {
+          this.ended = true;
+          this.emit('finish');
+        }
+        return;
+      }
+      this.doWrite(this.buffer.getBuffer());
+    };
+
+    if (Array.isArray(data)) {
+      if (this._writev) this._writev(data, cb);
+      else this._write(Buffer.concat(data), cb);
+    } else {
+      this._write(data, cb);
+    }
+    callbacks.forEach(cb => cb());
```

Comment:
This should be run in the `cb` as callback call means that the data has been written, not that it was passed to the underlying backend (flushed). (you can check out node docs or _stream_writable/doWrite & _stream_writable/onwrite).